File: library/src/main/java/org/xmtp/android/library/libxmtp/MessageV3.kt
package org.xmtp.android.library.libxmtp

import android.util.Log
import org.xmtp.android.library.Client
import org.xmtp.android.library.DecodedMessage
import org.xmtp.android.library.XMTPException
import org.xmtp.android.library.codecs.EncodedContent
import org.xmtp.android.library.messages.DecryptedMessage
import org.xmtp.android.library.messages.MessageDeliveryStatus
import org.xmtp.android.library.messages.Topic
import org.xmtp.android.library.toHex
import uniffi.xmtpv3.FfiDeliveryStatus
import uniffi.xmtpv3.FfiGroupMessageKind
import uniffi.xmtpv3.FfiMessage
import uniffi.xmtpv3.org.xmtp.android.library.codecs.ContentTypeGroupMembershipChange
import java.util.Date

data class MessageV3(val client: Client, private val libXMTPMessage: FfiMessage) {

    val id: ByteArray
        get() = libXMTPMessage.id

    val convoId: ByteArray
        get() = libXMTPMessage.convoId

    val senderAddress: String
        get() = libXMTPMessage.addrFrom

    val sentAt: Date
        get() = Date(libXMTPMessage.sentAtNs / 1_000_000)

    val deliveryStatus: MessageDeliveryStatus
        get() = when (libXMTPMessage.deliveryStatus) {
            FfiDeliveryStatus.UNPUBLISHED -> MessageDeliveryStatus.UNPUBLISHED
            FfiDeliveryStatus.PUBLISHED -> MessageDeliveryStatus.PUBLISHED
            FfiDeliveryStatus.FAILED -> MessageDeliveryStatus.FAILED
        }

    fun decode(): DecodedMessage {
        try {
            val decodedMessage = DecodedMessage(
                id = id.toHex(),
                client = client,
                topic = Topic.groupMessage(convoId.toHex()).description,
                encodedContent = EncodedContent.parseFrom(libXMTPMessage.content),
                senderAddress = senderAddress,
                sent = sentAt,
                deliveryStatus = deliveryStatus
            )
            if (decodedMessage.encodedContent.type == ContentTypeGroupMembershipChange && libXMTPMessage.kind != FfiGroupMessageKind.MEMBERSHIP_CHANGE) {
                throw XMTPException("Error decoding group membership change")
            }
            return decodedMessage
        } catch (e: Exception) {
            throw XMTPException("Error decoding message", e)
        }
    }

    fun decodeOrNull(): DecodedMessage? {
        return try {
            decode()
        } catch (e: Exception) {
            Log.d("MESSAGE_V3", "discarding message that failed to decode", e)
            null
        }
    }

    fun decryptOrNull(): DecryptedMessage? {
        return try {
            decrypt()
        } catch (e: Exception) {
            Log.d("MESSAGE_V3", "discarding message that failed to decrypt", e)
            null
        }
    }

    fun decrypt(): DecryptedMessage {
        return DecryptedMessage(
            id = id.toHex(),
            topic = Topic.groupMessage(convoId.toHex()).description,
            encodedContent = decode().encodedContent,
            senderAddress = senderAddress,
            sentAt = Date(),
            deliveryStatus = deliveryStatus
        )
    }
}

File: Sources/XMTPiOS/Mls/MessageV3.swift
//
//  MessageV3.swift
//  
//
//  Created by Naomi Plasterer on 4/10/24.
//

import Foundation
import LibXMTP

enum MessageV3Error: Error {
	case decodeError(String)
}

public struct MessageV3: Identifiable {
	let client: Client
	let ffiMessage: FfiMessage
	
	init(client: Client, ffiMessage: FfiMessage) {
		self.client = client
		self.ffiMessage = ffiMessage
	}
	
	public var id: Data {
		return ffiMessage.id
	}
	
	var convoId: Data {
		return ffiMessage.convoId
	}
	
	var senderAddress: String {
		return ffiMessage.addrFrom
	}
	
	var sentAt: Date {
		return Date(timeIntervalSince1970: TimeInterval(ffiMessage.sentAtNs) / 1_000_000_000)
	}
	
	var deliveryStatus: MessageDeliveryStatus {
		switch ffiMessage.deliveryStatus {
		case .unpublished:
			return .unpublished
		case .published:
			return .published
		case .failed:
			return .failed
		}
	}
	
	public func decode() throws -> DecodedMessage {
		do {
			let encodedContent = try EncodedContent(serializedData: ffiMessage.content)

			let decodedMessage = DecodedMessage(
				id: id.toHex,
				client: client,
				topic: Topic.groupMessage(convoId.toHex).description,
				encodedContent: encodedContent,
				senderAddress: senderAddress,
				sent: sentAt,
				deliveryStatus: deliveryStatus
			)
			
			if decodedMessage.encodedContent.type == ContentTypeGroupMembershipChanged && ffiMessage.kind != .membershipChange {
				throw MessageV3Error.decodeError("Error decoding group membership change")
		 }
			
			return decodedMessage
		} catch {
			throw MessageV3Error.decodeError("Error decoding message: \(error.localizedDescription)")
		}
	}
	
	public func decodeOrNull() -> DecodedMessage? {
		do {
			return try decode()
		} catch {
			print("MESSAGE_V3: discarding message that failed to decode", error)
			return nil
		}
	}
	
	public func decryptOrNull() -> DecryptedMessage? {
		do {
			return try decrypt()
		} catch {
			print("MESSAGE_V3: discarding message that failed to decrypt", error)
			return nil
		}
	}
	
	public func decrypt() throws -> DecryptedMessage {
		let encodedContent = try EncodedContent(serializedData: ffiMessage.content)

		let decrytedMessage =  DecryptedMessage(
			id: id.toHex,
			encodedContent: encodedContent,
			senderAddress: senderAddress,
			sentAt: Date(),
			topic: Topic.groupMessage(convoId.toHex).description,
			deliveryStatus: deliveryStatus
		)
		
		if decrytedMessage.encodedContent.type == ContentTypeGroupMembershipChanged && ffiMessage.kind != .membershipChange {
			throw MessageV3Error.decodeError("Error decoding group membership change")
		}
		
		return decrytedMessage
	}
}
